infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/billing/AzureBillingResourceUsage.java

```diff
@@ -196,69 +204,156 @@ String getLastResourceUsageHistory(AzureBillingCredentials azureBillingCredentia
+
+                // Which quantity to consider in this interval?
+                if (resourceQuantityInThatHour >= upperStepQuantity) {
+                    quantityToPriceInThisStep = upperStepQuantity - lowerStepQuantity;
+                } else if (resourceQuantityInThatHour > lowerStepQuantity) {
+                    quantityToPriceInThisStep = resourceQuantityInThatHour - lowerStepQuantity;
+                } else {
+                    break;
+                }
+                // Price it and add it to the global cost
+                costInThatHour += quantityToPriceInThisStep * lowerStepRate;
+                LOGGER.debug("AzureBillingResourceUsage computeCostInThatHour added to costInThatHour: " +
+                             quantityToPriceInThisStep + " x " + lowerStepRate + " [new costInThatHour = " +
+                             costInThatHour + "]");
+            }
+
+            // Update lowerStepQuantity & lowerStepRate
+            lowerStepQuantity = upperStepQuantity;
+            lowerStepRate = upperStepRate;
+        }
 
-        // Get the last resources usage history
-        String resourceUsageHistory = getLastResourceUsageHistory(azureBillingCredentials);
+        // The last quantity (without upperStepQuantity)
+        if (resourceQuantityInThatHour > lowerStepQuantity) {
+            quantityToPriceInThisStep = resourceQuantityInThatHour - lowerStepQuantity;
+            costInThatHour += quantityToPriceInThisStep * lowerStepRate;
 
-        // No available resource usage history
-        if (resourceUsageHistory == null)
-            return false;
+            LOGGER.debug("AzureBillingResourceUsage computeCostInThatHour last added to costInThatHour: " +
+                         quantityToPriceInThisStep + " x " + lowerStepRate + " [new costInThatHour = " +
+                         costInThatHour + "]");
+        }
 
-        // Parse resourceUsageHistory to find the desired resource usage
-        Iterator<JsonElement> resourceUsageIterator = JSON_PARSER.parse(resourceUsageHistory)
-                                                                 .getAsJsonObject()
-                                                                 .get("value")
-                                                                 .getAsJsonArray()
-                                                                 .iterator();
+        return costInThatHour;
+    }
 
-        while (resourceUsageIterator.hasNext()) {
-            JsonElement resourceUsage = resourceUsageIterator.next();
-            JsonObject resourceProperties = resourceUsage.getAsJsonObject().get("properties").getAsJsonObject();
+    // synchronized to ensure we dont try to use meter ids to get the meter rates while we are updating them
+    synchronized public HashSet<String> updateResourceUsageOrGetMetersIds(
+            AzureBillingCredentials azureBillingCredentials, HashMap<String, LinkedHashMap<String, Double>> metersRates,
```

Comment:
same remark of splitting method as above.

Reply:
(cf my previous comment)